sdk/core/azure-core/src/main/java/com/azure/core/implementation/TypeUtil.java

```diff
@@ -75,46 +79,52 @@ public static Type getTypeArgument(Type type) {
-            ParameterizedType parameterizedType = (ParameterizedType) type;
-            Type genericSuperClass = ((Class<?>) parameterizedType.getRawType()).getGenericSuperclass();
-            if (genericSuperClass instanceof ParameterizedType) {
-                /*
-                 * Find erased generic types for the super class and replace
-                 * with actual type arguments from the parameterized type
-                 */
-                Type[] superTypeArguments = getTypeArguments(genericSuperClass);
-                List<Type> typeParameters =
-                    Arrays.asList(((Class<?>) parameterizedType.getRawType()).getTypeParameters());
-                int j = 0;
-                for (int i = 0; i != superTypeArguments.length; i++) {
-                    if (typeParameters.contains(superTypeArguments[i])) {
-                        superTypeArguments[i] = parameterizedType.getActualTypeArguments()[j++];
+    public static Type getSuperType(final Type type) {
+        return SUPER_TYPE_MAP.computeIfAbsent(type, _type -> {
+            if (type instanceof ParameterizedType) {
+                ParameterizedType parameterizedType = (ParameterizedType) type;
+                Type genericSuperClass = ((Class<?>) parameterizedType.getRawType()).getGenericSuperclass();
+                if (genericSuperClass instanceof ParameterizedType) {
+                    /*
+                     * Find erased generic types for the super class and replace
+                     * with actual type arguments from the parameterized type
+                     */
+                    final Type[] superTypeArguments = getTypeArguments(genericSuperClass);
+                    final Type[] typeParameters =
+                        ((GenericDeclaration) parameterizedType.getRawType()).getTypeParameters();
+                    int k = 0;
+                    outer:
+                    for (int i = 0; i != superTypeArguments.length; i++) {
+                        for (int j = 0; i < typeParameters.length; j++) {
+                            if (typeParameters[j].equals(superTypeArguments[i])) {
+                                superTypeArguments[i] = parameterizedType.getActualTypeArguments()[k++];
+                                break outer;
+                            }
+                        }
                     }
+                    return new ParameterizedType() {
+                        @Override
+                        public Type[] getActualTypeArguments() {
+                            return superTypeArguments;
+                        }
+
+                        @Override
+                        public Type getRawType() {
+                            return ((ParameterizedType) genericSuperClass).getRawType();
+                        }
+
+                        @Override
+                        public Type getOwnerType() {
+                            return null;
+                        }
+                    };
```

Comment:
Could this call `createParameterizedType`?

Reply:
Yep, done